src/app/[locale]/specs/sync/page.mdx

```diff
@@ -0,0 +1,171 @@
+
+## Bootstrapping a Live Mirror
+
+The firehose can be used to follow new data updates, and repo exports can be used for snapshots. Actually combining the two to bootstrap a complete live-updating mirror can be a bit tricky. One approach is described below.
+
+Keep a sync status table for all accounts (DIDs) encountered. The status can be:
+
+- `dirty`: there is either no local repo data for this account, or it has gotten out of sync
+- `in-process`: the repo is "dirty", but there is a background task in process to update it
+- `synchronized`: a complete copy of the repository has been processed
+
+Start by subscribing to the full firehose. If there is no existing repository data for the account, mark the account as "dirty". When new events come in for a repo, the behavior depends on the repo state. If it is "dirty", the event is ignored. If the state is "synchronized", the event is immediately processed as an update to the repo. If the state is "in-process", the event is enqueued locally.
+
+Have a set of background workers start processing "dirty" repos. First they mark the status as `in-process`, so that new events are enqueued locally. Then the full repo export (CAR file) is fetched from the PDS and processed in full. The commit `rev` of the repo export is noted. When the full repo import is complete, the worker can start processing any enqueued events, in order, skipping any with a `rev` lower than the existing repo processed `rev` (as is the usual behavior). When the queue for the account is fully processed, the state can be flipped to `synchronized`, and the worker can move on.
+
+After some time, most of the known accounts will be marked as `synchronized`, though this will only represent the most recently active accounts in the network. Next a more complete set of repositories in the network can be fetched, for example using an API query against an existing large service. Any new identified accounts can be marked as `dirty` in the service, and the background workers can start processing them.
+
+When all of the accounts are `synchronized`, the process is complete. At large scale it may be hard to get perfect synchronization: PDS instances may be down at various times, identities may fail to resolve, or invalid events, data, or signatures may end up in the network.
+
+## Security Concerns
+
+General mitigations for resource exhaustion attacks are recommended: event rate-limits, data quotas per account, limits on data object sizes and deserialized data complexity, etc.
+
+Care should always be taken when making network requests to unknown or untrusted hosts, especially when the network locators for those host from from untrusted input. This includes validating URLs to not connect to local or internal hosts (including via HTTP redirects), avoiding SSRF in browser contexts, etc.
+
```

Comment:
Might want to draw specific attention to being mindful of *outbound* request rates.

Consider multiple PDSes colluding to spam `#identity` events, each below the inbound event rate limits for an individual PDS, but collectively driving lots of handle-resolution traffic towards hosts of their choice.

Reply:
This might even be relevant for non-malicious traffic, I vaguely recall someone talking about getting blocked by their DNS resolver for trying to resolve too many handles.

Reply:
added a note